### ISSUE TITLE
Online Clustering

### ISSUE THREAD

**Issue by ShyamGanesh13** [MaartenGr/BERTopic#1790]
Hi @MaartenGr , 
I am trying to use Online topic modelling to solve my use-case. I am using RIVER(DBSTREAM) as mentioned in your help docs.
This DBSTREAM model of river package is a density-based algorithm, but it is not helping in finding the outliers in my dataset.
Assuming that my dataset has outlier, Is there any solution through which I can get outliers as well while using the above algorithm?

**Comment by MaartenGr** [comment]
I would advise using the newly introduced [`.merge_models`](https://maartengr.github.io/BERTopic/getting_started/merge/merge.html) technique instead. It allows you to combine multiple models to essentially have an online-learning approach.

**Comment by yamsgithub** [comment]
I have a large number of documents. About 1 million. So I wanted to use online clustering especially as I want to also calculate the topic distributions.

However, online topic modeling is not working. I get only one single topic assigned for about a million short docs. Which is also evident from the topic distributions which all have a single value. My code is as follows:

`from river import stream
from river import cluster

class River:
    def __init__(self, model):
        self.model = model

    def partial_fit(self, umap_embeddings):
        for umap_embedding, _ in stream.iter_array(umap_embeddings):
            self.model.learn_one(umap_embedding)

        labels = []
        for umap_embedding, _ in stream.iter_array(umap_embeddings):
            label = self.model.predict_one(umap_embedding)
            labels.append(label)

        self.labels_ = labels
        return self`

`from sklearn.cluster import MiniBatchKMeans
from sklearn.decomposition import IncrementalPCA
from bertopic.vectorizers import OnlineCountVectorizer
from bertopic import BERTopic
from bertopic.representation import MaximalMarginalRelevance, OpenAI
from sentence_transformers import SentenceTransformer
import openai
from bertopic.dimensionality import BaseDimensionalityReduction


min_topic_size = 5
nr_topics = None
calculate_probabilities = True
embedding_model = SentenceTransformer("all-MiniLM-L6-v2")
umap_model = IncrementalPCA(n_components=5)
# umap_model = BaseDimensionalityReduction()
# cluster_model = MiniBatchKMeans(n_clusters=50, random_state=0)
cluster_model = River(cluster.DBSTREAM())
vectorizer_model = OnlineCountVectorizer(stop_words="english", decay=.01)

api_key = 'sk-proj-0qmfLabg9c4DQRfL366wT3BlbkFJ262pDtGOZcmPXQcrXnhZ'
client = openai.OpenAI(api_key=api_key)
openai_generator = OpenAI(client, model="gpt-4o-mini", delay_in_seconds=1)
mmr = MaximalMarginalRelevance(diversity=0.3)
representation_models = [mmr, openai_generator]

topic_model = BERTopic(representation_model=representation_models, 
                       embedding_model=embedding_model,                        
                       min_topic_size=min_topic_size, 
                       nr_topics=nr_topics, 
                       calculate_probabilities=calculate_probabilities, 
                       umap_model=umap_model,
                       hdbscan_model=cluster_model,
                       vectorizer_model=vectorizer_model,
                       verbose=True)

topics = []
topic_distrs = []
doc_count = 0
doc_chunk = 10000
total_doc_count = len(df)
# total_doc_count = 500000
while doc_count < total_doc_count:
  docs = df['body'][doc_count:doc_count+doc_chunk]
  topic_model.partial_fit(docs)
  topics.extend(topic_model.topics_)
  doc_count += (doc_chunk)

topic_model.topics_ = topics

topic_model.save(os.path.join(MODEL_PATH,'np_dc_online_topic_model_subset'), serialization="safetensors")

topic_distrs, _ = topic_model.approximate_distribution(df_pt['body'][:total_doc_count])
`

So topic_distrs is:

array([[1.],
       [1.],
       [1.],
       ...,
       [1.],
       [1.],
       [1.]])

Am I doing something incorrect? I followed all the intructions from https://maartengr.github.io/BERTopic/getting_started/online/online.html

**Comment by yamsgithub** [comment]
I also tried using merge models but then I do not seem to be able to get topic distributions after I merge the models. Is there a way to get the topic distributions of the merged model for all the documents?

**Comment by yamsgithub** [comment]
I have a large number of documents. About 1 million. So I wanted to use online clustering especially as I want to also calculate the topic distributions.

However, online topic modeling is not working. I get only one single topic assigned for about a million short docs. Which is also evident from the topic distributions which all have a single value. My code is as follows:

`from river import stream
from river import cluster

class River:
    def __init__(self, model):
        self.model = model

    def partial_fit(self, umap_embeddings):
        for umap_embedding, _ in stream.iter_array(umap_embeddings):
            self.model.learn_one(umap_embedding)

        labels = []
        for umap_embedding, _ in stream.iter_array(umap_embeddings):
            label = self.model.predict_one(umap_embedding)
            labels.append(label)

        self.labels_ = labels
        return self`

`from sklearn.cluster import MiniBatchKMeans
from sklearn.decomposition import IncrementalPCA
from bertopic.vectorizers import OnlineCountVectorizer
from bertopic import BERTopic
from bertopic.representation import MaximalMarginalRelevance, OpenAI
from sentence_transformers import SentenceTransformer
import openai
from bertopic.dimensionality import BaseDimensionalityReduction


min_topic_size = 5
nr_topics = None
calculate_probabilities = True
embedding_model = SentenceTransformer("all-MiniLM-L6-v2")
umap_model = IncrementalPCA(n_components=5)
cluster_model = River(cluster.DBSTREAM())
vectorizer_model = OnlineCountVectorizer(stop_words="english", decay=.01)

api_key = <api_key>
client = openai.OpenAI(api_key=api_key)
openai_generator = OpenAI(client, model="gpt-4o-mini", delay_in_seconds=1)
mmr = MaximalMarginalRelevance(diversity=0.3)
representation_models = [mmr, openai_generator]

topic_model = BERTopic(representation_model=representation_models, 
                       embedding_model=embedding_model,                        
                       min_topic_size=min_topic_size, 
                       nr_topics=nr_topics, 
                       calculate_probabilities=calculate_probabilities, 
                       umap_model=umap_model,
                       hdbscan_model=cluster_model,
                       vectorizer_model=vectorizer_model,
                       verbose=True)

topics = []
topic_distrs = []
doc_count = 0
doc_chunk = 10000
total_doc_count = len(df)
while doc_count < total_doc_count:
  docs = df['body'][doc_count:doc_count+doc_chunk]
  topic_model.partial_fit(docs)
  topics.extend(topic_model.topics_)
  doc_count += (doc_chunk)

topic_model.topics_ = topics

topic_model.save(os.path.join(MODEL_PATH,'np_dc_online_topic_model_subset'), serialization="safetensors")

topic_distrs, _ = topic_model.approximate_distribution(df_pt['body'][:total_doc_count])
`

So topic_distrs is:

array([[1.],
       [1.],
       [1.],
       ...,
       [1.],
       [1.],
       [1.]])

Am I doing something incorrect? I followed all the intructions from https://maartengr.github.io/BERTopic/getting_started/online/online.html
